### PR TITLE
moved _parse_species_list function to tardis.io.util.py 

### PR DESCRIPTION
**Pull request by rudrakshnalbalwar**
### :pencil: Description

**Type:** :roller_coaster: `infrastructure`

Relocated species list parsing logic from liv_plot.py to util.py to improve file organization and maintainability. No changes to functionality were made; the parsing logic remains the same but is now centralized in a more appropriate file.

### :pushpin: Resources

image

### :vertical_traffic_light: Testing

How did you test these changes?

- [ ] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)

### :ballot_box_with_check: Checklist

- [ ] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

### CODE REVIEW

File: tardis/io/util.py [hunk 1]
```diff
@@ -138,6 +138,51 @@ def yaml_load_file(filename, loader=yaml.Loader):
...
+        nelements : int, optional
+            Number of elements to include in plot. The most interacting elements are included. If None, displays all elements.
+
+        Raises
+        ------
+        ValueError
+            If species list contains invalid entries.
+
+        """
+        from tardis.util.base import atomic_number2element_symbol
```
Comment (atharva-2001) [inline tardis/io/util.py:162]: I don't think this should be here

File: tardis/io/util.py [hunk 2]
```diff
@@ -138,6 +138,51 @@ def yaml_load_file(filename, loader=yaml.Loader):
         return yaml.load(stream, Loader=loader)
 
 
+def _parse_species_list(self, species_list, packets_mode, nelements=None):
```
Comment (atharva-2001) [inline tardis/io/util.py:141]: no need for the underscore I think

Comment (rudrakshnalbalwar) [reply tardis/io/util.py]: done the changes as you mentioned

**Comment by jvshields** [comment]
Thanks for the contribution! This is a good change, but I think you need to double check where this function is used in the codebase, and change any imports or function calls to point to the right location. I believe that this function is used in the LIV and SDEC plot classes. https://github.com/search?q=repo%3Atardis-sn%2Ftardis%20_parse_species_list&type=code

**Comment by rudrakshnalbalwar** [comment]
@jvshields  I made the changes as you mentioned , I’ve removed the underscore from the parse_species_list function name as requested. However, I did not change the name of the test function test_parse_species_list, since it already follows the convention for testing this function. And I noticed that there were no explicit imports for the _parse_species_list function previously. Since the function is used within the same class/module, an import doesn't seem necessary in this case. Therefore, I haven't added an import for parse_species_list either.

**Comment by rudrakshnalbalwar** [comment]
Also one more thing do I have to make changes here too ?
tardis/.pytest_cache/v/cache/nodeids
@jvshields , @atharva-2001 , @andrewfullard

File: tardis/visualization/tools/sdec_plot.py [hunk 3]
```diff
@@ -508,7 +508,7 @@ def from_hdf(cls, hdf_fpath, packets_mode=None):
                 }
             )
 
-    def _parse_species_list(self, species_list):
+    def parse_species_list(self, species_list):
```
Comment (atharva-2001) [inline tardis/visualization/tools/sdec_plot.py:511]: Is this being imported?

File: tardis/visualization/tools/tests/test_sdec_plot.py [hunk 4]
```diff
@@ -172,7 +172,7 @@ def test_parse_species_list(self, request, plotter, attribute):
         species : list
         """
         # THIS NEEDS TO BE RUN FIRST. NOT INDEPENDENT TESTS
-        plotter._parse_species_list(self.species_list[0])
+        plotter.parse_species_list(self.species_list[0])
```
Comment (atharva-2001) [inline tardis/visualization/tools/tests/test_sdec_plot.py:175]: this needs to be tested elsewhere

**Comment by jvshields** [comment]
Please double check that the tests are passing. 

I believe that the full refactor here should look like this. 

First, move the parse_species_list() function to the util file. (You have done this and it looks good)

Next, remove the function definition from the sdec plot and liv plot classes. Import the more generic function from the util file.

Finally, wherever the function is used in those classes, make sure that it's working properly with the new imported function.
The tests here will tell you if everything is hooked up correctly. 

Thanks.

File: tardis/visualization/tools/sdec_plot.py (hunk 3 above)
Comment (rudrakshnalbalwar) [reply tardis/visualization/tools/sdec_plot.py:511]: yes

**Comment by rudrakshnalbalwar** [comment]
> Please double check that the tests are passing.
> 
> I believe that the full refactor here should look like this.
> 
> First, move the parse_species_list() function to the util file. (You have done this and it looks good)
> 
> Next, remove the function definition from the sdec plot and liv plot classes. Import the more generic function from the util file.
> 
> Finally, wherever the function is used in those classes, make sure that it's working properly with the new imported function. The tests here will tell you if everything is hooked up correctly.
> 
> Thanks.

As you mentioned, I've implemented those changes, and I ran all the tests locally on my device. They are all passing successfully. Could you please review the changes?

File: tardis/io/util.py [hunk 5]
```diff
@@ -138,6 +138,57 @@ def yaml_load_file(filename, loader=yaml.Loader):
...
+        nelements : int, optional
+            Number of elements to include in plot. The most interacting elements are included. If None, displays all elements.
+
+        Raises
+        ------
+        ValueError
+            If species list contains invalid entries.
+
+        """
+        from tardis.util.base import atomic_number2element_symbol
```
Comment (atharva-2001) [inline tardis/io/util.py:162]: why is this import defined here?

File: tardis/visualization/tools/sdec_plot.py (hunk 3 above)
Comment (atharva-2001) [reply tardis/visualization/tools/sdec_plot.py:511]: does this serve a different functionality?

File: tardis/io/util.py [hunk 6]
```diff
@@ -160,7 +211,7 @@ def traverse_configs(base, other, func, *args):
             traverse_configs(base[k], other[k], func, *args)
     elif (
         isinstance(base, collections_abc.Iterable)
-        and not isinstance(base, basestring)
```
Comment (wkerzendorf) [inline tardis/io/util.py:163]: why is this changed?

File: tardis/io/util.py [hunk 7]
```diff
@@ -138,6 +138,57 @@ def yaml_load_file(filename, loader=yaml.Loader):
         return yaml.load(stream, Loader=loader)
 
 
+def parse_species_list(sdec_plotter, data, species_list, packets_mode, nelements=None):
```
Comment (wkerzendorf) [inline tardis/io/util.py:141]: why does it take a plotter argument? What is the purpose of this function?

File: tardis/io/util.py (hunk 5 above)
Comment (rudrakshnalbalwar) [reply tardis/io/util.py:162]: It is defined inside the function to break the circular dependency between tardis.io.util and tardis.util.base

Comment (rudrakshnalbalwar) [reply tardis/io/util.py:162]: and in the below sdec_plot
The parse_species_list function in util.py builds on the sdec_plot.py version by not only parsing species but also filtering them based on interaction frequency in the data. This makes it more data-driven and adaptable for cases where top interacting elements are needed.

File: tardis/io/util.py (hunk 6 above)
Comment (rudrakshnalbalwar) [reply tardis/io/util.py:163]: the change was made to ensure that function is compatible with python 3

File: tardis/io/util.py (hunk 7 above)
Comment (rudrakshnalbalwar) [reply tardis/io/util.py:141]: because it relies on some methods and attributes within sdec_plotter to parse and organize the species list.

File: tardis/io/util.py (hunk 5 above)
Comment (atharva-2001) [reply tardis/io/util.py:162]: Imports should be defined at the start of the file right?

Comment (rudrakshnalbalwar) [reply tardis/io/util.py:162]: Yes, but importing this statement at the start of the file creates a circular dependency. Specifically, atomic_number2element_symbol is imported from tardis.util.base, and within tardis.util.base, get_internal_data_path is imported from util.py, which triggers a circular loop. Refactoring the functions to avoid this would increase code complexity and could introduce new errors.

**Comment by rudrakshnalbalwar** [comment]
Hey @atharva-2001 and Prof. @wkerzendorf,
I've made the necessary changes to both sdec_plot.py and test_sdec_plot.py files as part of refactoring the parse_species_list functionality. The function is now correctly imported from util.py after being moved from liv_plot.py.
Could you please review the changes and let me know if everything looks good or if any further adjustments are needed?

**Comment by andrewfullard** [comment]
Thanks for the contribution. Unfortunately we have decided to take a different path to refactor this part of the code. Specifically your plan to reference the `sdec_plotter` object within the function makes it less useful as a general utility function.
